commons/zenoh-buffers/src/slice.rs

```diff
@@ -165,39 +133,30 @@ impl<'a> HasReader for &'a [u8] {
 
 impl Reader for &[u8] {
     fn read(&mut self, into: &mut [u8]) -> Result<NonZeroUsize, DidntRead> {
-        let len = self.len().min(into.len());
-        // SAFETY: len is guaranteed to be the smaller than lhs length.
-        let lhs = crate::unsafe_slice_mut!(into, ..len);
-        let rhs = crate::unsafe_slice!(self, ..len);
-        lhs.copy_from_slice(rhs);
-        // SAFETY: len is guaranteed to be smaller than slice.len().
-        *self = crate::unsafe_slice!(self, len..);
-        NonZeroUsize::new(len).ok_or(DidntRead)
+        let Some(len) = NonZeroUsize::new(self.len().min(into.len())) else {
+            return Err(DidntRead);
+        };
+        let (to_write, remain) = self.split_at(len.get());
+        into[..len.get()].copy_from_slice(to_write);
+        *self = remain;
+        Ok(len)
     }
 
     fn read_exact(&mut self, into: &mut [u8]) -> Result<(), DidntRead> {
         let len = into.len();
         if self.len() < len {
             return Err(DidntRead);
         }
-        // SAFETY: len is guaranteed to be the smaller than lhs length.
-        let lhs = crate::unsafe_slice_mut!(into, ..len);
-        let rhs = crate::unsafe_slice!(self, ..len);
-        lhs.copy_from_slice(rhs);
-        // SAFETY: len is guaranteed to be smaller than slice.len().
-        *self = crate::unsafe_slice!(self, len..);
+        let (to_write, remain) = self.split_at(len);
+        into[..len].copy_from_slice(to_write);
```

Comment:
In general you are re-introducing a lot of *slice boundaries* checks that have an impact at runtime while guarantees are held by the contract and covered by tests. Since these operations are very hot in the code, impact is not negligible and assessed in the past on few `100k msg/s`.

Reply:
I've checked before on Godbolt than bound checks are elided. See the links in the PR description. If you're not satisfied, I can also run `cargo asm` to ensure that bound checks are removed.

Reply:
Btw, we are already using elided checks, because `copy_from_slice` have an assertion about the slices' length being equal. So to be consistent with this full unsafe approach, we should directly use `ptr::copy_nonoverlapping` instead of `copy_from_slice`.